### ISSUE TITLE
Context Parallel for Qwen3

### ISSUE THREAD

**Issue by unavailableun** [pytorch/torchtitan#2077]
Thanks for supporting Qwen3 models!

> CP is not supported currently because of RoPE embedding implementation details.

Any plan to support CP + EP for Qwen3 MoE models?
If no plan in short time, can you help guide how can I implement it myself?

**Comment by tianyu-l** [comment]
cc @fegin

**Comment by unavailableun** [comment]
I saw alert only for CP & FlexAttention, does that mean CP works with SDPA?
<img width="1170" height="492" alt="Image" src="https://github.com/user-attachments/assets/d03cf3b7-3dba-4804-9369-0ef5da98f636" />

**Comment by tianyu-l** [comment]
cc @wwwjn @shuhuayu

**Comment by wwwjn** [comment]
Thanks for asking! Now Qwen3 is not officially supported for Qwen3, and we haven't implemented and tested CP on Qwen3.  This is because of the RoPE embedding differences (In Qwen3, the RoPE field is named "rope_cache" while in `train.py`, it's looking for `freqs_cis` field. 
Qwen3 README: https://github.com/pytorch/torchtitan/tree/refs/heads/main/torchtitan/models/qwen3
Train.py: https://github.com/pytorch/torchtitan/blob/refs/heads/main/torchtitan/train.py#L480

So it won't work with SDPA, but it should be a simple change

**Comment by pstefa1707** [comment]
What I did to get Qwen3 working with CP:

1. Added an alias property to the Qwen3 model class for `freqs_cis` -> `rope_cache`
```
    @property
    def freqs_cis(self) -> torch.Tensor:
        """Alias for rope_cache to maintain compatibility with torchtitan context parallelism."""
        return self.rope_cache
```

2. Ran into some bugs in the cuDNN backend for SDPA so excluded it from `models.attention.ScaledDotProductAttentionWrapper`

I also ran into a deeper memory leak bug related to running EP and FSDP (for the MoE variants), where it would cause us to accumulate memory during the backwards pass and OOM.

This was solved by updating the the explicit prefetch lists set in [llama4/infra/parallelize.py#L425](https://github.com/pytorch/torchtitan/blob/fbafd44da2baef0afac58989f07d799c4251bdef/torchtitan/models/llama4/infra/parallelize.py#L425):

```
# layer_N should prefetch its OWN experts (the next thing to run)
transformer_block.set_modules_to_backward_prefetch([transformer_block.moe.experts])
# layer_N.experts should prefetch layer_{N-1} (the next layer)
if prev_transformer_block is not None:
      transformer_block.moe.experts.set_modules_to_backward_prefetch(
              [prev_transformer_block]
      )
elif model.tok_embeddings is not None:
      transformer_block.moe.experts.set_modules_to_backward_prefetch(
              [model.tok_embeddings]
      )
else:
      transformer_block.moe.experts.set_modules_to_backward_prefetch([])
```

This above issue is unrelated but was a blocker. I hope to provide a write up of my understanding of the issue and this solution (and a reprod example) but haven't had the time yet.

**Comment by tianyu-l** [comment]
@shuhuayu could you please take a look at the memory leak issue mentioned above?

**Comment by pstefa1707** [comment]
I have tried to create a minimal reproducible example but spent an hour with the debug sized model but couldn't get it (I was only facing this getting qwen3-235b training on 2 nodes). I also had some local changes to qwen3 for adding LoRAs, although I don't think they'd impact this.

I appreciate this is not a useful bug report. If this is something anyone else has run into I am keen to spent more time to package this bug up into something that can be iterated on.

Here are the before and after fix memory snapshots. Note the accumulating memory on the backwards pass.

<img width="5112" height="1626" alt="Image" src="https://github.com/user-attachments/assets/b624e6a2-606c-4eb7-ba67-c777a14ddb31" />
<img width="2564" height="812" alt="Image" src="https://github.com/user-attachments/assets/6c5d713b-163a-4055-8e1c-2ff40fdafc5f" />

Here are my personal debug notes, forgive the vibey-ness. I am happy to hop on a call and discuss if needed. My thesis may very well be wrong.

<h2>Summary</h2>
<p>A memory leak during the backward pass of training Qwen3-235B-A22B, where <strong>~0.5GB of GPU memory accumulated per layer</strong> processed. With 94 layers, this led to OOM well before backward pass completion. The root cause was <strong>FSDP2's implicit backward prefetch mechanism fetching weights for layers that had already completed their backward pass</strong>, resulting in unsharded tensors that were never freed.</p>
<hr>
<h2>The Architecture</h2>
<h3>FSDP2 Module Hierarchy</h3>
<p>The model has a nested FSDP structure:</p>
<pre><code>Model (FSDPModule)
├── tok_embeddings (FSDPModule)
├── layers.0 (FSDPModule - TransformerBlock)
│   └── moe.experts (FSDPModule - separate mesh!)  ← Key: This is ALSO wrapped
├── layers.1 (FSDPModule - TransformerBlock)
│   └── moe.experts (FSDPModule)
├── ...
├── layers.93 (FSDPModule - TransformerBlock)
│   └── moe.experts (FSDPModule)
└── norm, output (FSDPModule)

</code></pre>
<p>Each <code>TransformerBlock</code> AND its inner <code>moe.experts</code> are <strong>independently FSDP-wrapped</strong>. The experts are on a different mesh (<code>dp_mod_ep_mesh</code>) due to Expert Parallelism.</p>
<h3>Two Levels of Prefetching in FSDP2</h3>
<p>FSDP2 has <strong>two distinct prefetch mechanisms</strong>:</p>
<ol>
<li><strong>Module-level prefetch</strong> (explicit API):
<ul>
<li>Controlled via <code>module.set_modules_to_backward_prefetch([...])</code></li>
<li>Stored in <code>_modules_to_backward_prefetch</code> attribute</li>
<li>This is what torchtitan's <code>apply_fsdp</code> configures</li>
</ul>
</li>
<li><strong>ParamGroup-level prefetch</strong> (implicit/internal):
<ul>
<li>Controlled internally by <code>FSDPParamGroup._backward_prefetch()</code></li>
<li>Called from <code>FSDPParamGroup.pre_backward()</code> hook</li>
<li><strong>Not exposed via any public API</strong></li>
<li>Uses internal logic to decide what to prefetch</li>
</ul>
</li>
</ol>
<hr>
<h2>The Bug: Detailed Mechanism</h2>
<h3>Normal FSDP2 Backward Flow</h3>
<p>For a single layer's backward pass, the correct flow is:</p>
<pre><code>1. pre_backward hook fires
2. UNSHARD layer N's weights (all-gather)
3. Compute gradients for layer N
4. post_backward hook fires
5. RESHARD layer N's weights (reduce-scatter gradients, free unsharded weights)
6. Move to layer N-1

</code></pre>
<h3>What Was Actually Happening</h3>
<p>With activation checkpointing (<code>mode = &quot;full&quot;</code>) and nested FSDP for MoE experts:</p>
<pre><code>Processing Layer 34's backward:
─────────────────────────────────────────────────────────
1. UNSHARD layers.34 (TransformerBlock wrapper)              [+139MB]
2. UNSHARD layers.34.moe.experts (param group 1)             [+562MB]
3. UNSHARD layers.34.moe.experts (param group 2)             [+562MB]
   ↑ These are for AC forward recomputation
4. RESHARD layers.34.moe.experts (AC forward done)           [-1.1GB]
5. UNSHARD layers.34.moe.experts (for actual backward)       [+562MB]
6. ⚠️ UNSHARD layers.35.moe.experts ← IMPLICIT PREFETCH      [+562MB]
7. POST_BACKWARD layers.34.moe.experts
8. RESHARD layers.34.moe.experts                             [-562MB]
9. POST_BACKWARD layers.34
10. RESHARD layers.34                                        [-139MB]

After layer 34 complete: layers.35.moe.experts STILL UNSHARDED!

</code></pre>
<h3>Why Layer 35's Experts Were Never Freed</h3>
<p>The critical issue at step 6:</p>
<ul>
<li><code>FSDPParamGroup._backward_prefetch()</code> for layer 34's experts was triggered</li>
<li>This internal method decided to prefetch <strong>layer 35's experts</strong></li>
<li>But layer 35's backward pass had <strong>already completed</strong> (we're going 93→0)</li>
<li>Since layer 35's <code>post_backward</code> hook already ran, there was no subsequent hook to RESHARD those prefetched weights</li>
<li><strong>The ~562MB stayed allocated indefinitely</strong></li>
</ul>
<h3>The Accumulation Pattern</h3>
<pre><code>Layer 93 backward complete: baseline memory
Layer 92 backward complete: +0.5GB (layer 93 experts prefetched, never freed)
Layer 91 backward complete: +0.5GB (layer 92 experts prefetched, never freed)
...
Layer 34 backward complete: +30GB cumulative
Layer 33 backward complete: +30.5GB cumulative
...
Layer ~27: OOM at ~71GB

</code></pre>
<hr>
<h2>Why Explicit Prefetch Settings Didn't Work</h2>
<h3>What torchtitan's <code>set_modules_to_backward_prefetch([])</code> Does</h3>
<p>This sets the <strong>module-level</strong> prefetch list to empty:</p>
<pre><code class="language-python">transformer_block._modules_to_backward_prefetch = []
transformer_block.moe.experts._modules_to_backward_prefetch = []

</code></pre>
<h3>What It Doesn't Do</h3>
<p>It does <strong>NOT</strong> affect <code>FSDPParamGroup._backward_prefetch()</code>, which is an <strong>internal method</strong> that:</p>
<ol>
<li>Is called from the <code>pre_backward</code> hook on the param group</li>
<li>Has its own logic for determining what to prefetch</li>
<li>Appears to prefetch based on <strong>module registration order</strong> (forward order: 0→93)</li>
<li>Ignores the module-level <code>_modules_to_backward_prefetch</code> setting</li>
</ol>
<h3>The Two Prefetch Systems Don't Talk to Each Other</h3>
<pre><code>Module Level (public API):
    FSDPModule.set_modules_to_backward_prefetch([])
    → Sets FSDPModule._modules_to_backward_prefetch = []
    → Affects nothing in your case (was already None)

ParamGroup Level (internal):
    FSDPParamGroup.pre_backward()
    → Calls FSDPParamGroup._backward_prefetch()
    → Uses internal/implicit logic
    → Ignores module-level settings
    → THIS IS WHAT WAS CAUSING THE LEAK

</code></pre>
<hr>
<h2>Why the Prefetch Was Going in the Wrong Direction</h2>
<h3>The Root Cause: Module Registration Order</h3>
<p>When FSDP2 wraps modules, it records them in <strong>forward order</strong>:</p>
<ul>
<li><code>layers.0</code>, <code>layers.1</code>, ..., <code>layers.93</code></li>
</ul>
<p>The implicit prefetch in <code>FSDPParamGroup._backward_prefetch()</code> appears to use this registration order to determine prefetch targets, possibly via something like:</p>
<pre><code class="language-python"># Pseudo-code of what FSDP2 might be doing internally
def _backward_prefetch(self):
    next_module = self._get_next_registered_module()  # Based on forward order!
    if next_module:
        next_module.unshard()  # Prefetch for &quot;next&quot; module

</code></pre>
<p>But during backward, we process in <strong>reverse order</strong> (93→0). So when processing layer 34:</p>
<ul>
<li>FSDP2 thinks &quot;next&quot; is layer 35 (based on forward registration order)</li>
<li>But in backward, &quot;next&quot; should be layer 33</li>
<li>Layer 35 was already processed, so its <code>post_backward</code> won't run again</li>
<li>The prefetched weights are orphaned</li>
</ul>
<hr>
<h2>The Fix: Disabling Implicit Prefetch</h2>
<h3>What the Monkey-Patch Does</h3>
<pre><code class="language-python">def _disabled_backward_prefetch(self):
    pass  # Do nothing

fsdp_pg.FSDPParamGroup._backward_prefetch = _disabled_backward_prefetch

</code></pre>
<p>This completely bypasses the implicit prefetch logic at the <code>FSDPParamGroup</code> level. Now the backward pass flow becomes:</p>
<pre><code>Processing Layer 34's backward:
─────────────────────────────────────────────────────────
1. UNSHARD layers.34
2. UNSHARD layers.34.moe.experts (for AC recompute)
3. RESHARD layers.34.moe.experts (AC done)
4. UNSHARD layers.34.moe.experts (for backward)
5. POST_BACKWARD layers.34.moe.experts
6. RESHARD layers.34.moe.experts                    ← Properly freed!
7. POST_BACKWARD layers.34
8. RESHARD layers.34                                ← Properly freed!

After layer 34 complete: NO orphaned tensors!
</code></pre>

Attached is the memory snapshot for the OOM, has mem-allocation stack traces which may be useful for providing more context.
[235b_oom_always_reshard_full_ac.pickle.zip](https://github.com/user-attachments/files/24133045/235b_oom_always_reshard_full_ac.pickle.zip)

**Comment by fegin** [comment]
I tentatively enable CP + SDPA for Qwen3 in https://github.com/pytorch/torchtitan/pull/2144. But I haven't verified the EP + CP part, which we may need some verifications.

**Comment by wwwjn** [comment]
#2446

**Comment by acisseJZhong** [comment]
@fegin Would CP work with Flex attention? I am hitting `Context Parallel only supports SDPA attention. Got attn_backend='flex'. FlexAttention and varlen attention are not supported with CP.` I can't use sdpa since I am using customized mask in RL usecase, I can either use Flex or Varlen. 

> But I haven't verified the EP + CP part, which we may need some verifications.

I can also help test EP+CP+TP+FSDP through running titan in verl.

**Comment by tianyu-l** [comment]
@acisseJZhong I think TP+CP has issue which https://github.com/pytorch/torchtitan/pull/2480 seems trying to workaround.

**Comment by wwwjn** [comment]
@fegin  Can we close this issue because we have another issue to tracking #2446 , and #2480 as a temporary workaround?

**Comment by acisseJZhong** [comment]
@wwwjn  we can close it, @fegin already landed the PR to unblock CP for flex attention for qwen3.
